go-controller/pkg/ovn/base_network_controller_multipolicy.go

```diff
@@ -0,0 +1,123 @@
+package ovn
+
+import (
+	"fmt"
+	"strings"
+
+	multinetworkpolicyapi "github.com/k8snetworkplumbingwg/multi-networkpolicy/pkg/apis/k8s.cni.cncf.io/v1beta1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	knet "k8s.io/api/networking/v1"
+	"k8s.io/klog/v2"
+)
+
+const PolicyForAnnotation = "k8s.v1.cni.cncf.io/policy-for"
+
+func (bnc *BaseNetworkController) syncMultiNetworkPolicies(multiPolicies []interface{}) error {
+	expectedPolicies := make(map[string]map[string]bool)
+	for _, npInterface := range multiPolicies {
+		policy, ok := npInterface.(*multinetworkpolicyapi.MultiNetworkPolicy)
+		if !ok {
+			return fmt.Errorf("spurious object in syncMultiNetworkPolicies: %v", npInterface)
+		}
+		if !bnc.shouldApplyMultiPolicy(policy) {
+			klog.V(5).Infof("[network (%s)] skipping syncing policy %s/%s",
+				bnc.GetNetworkName(), policy.Namespace, policy.Name)
+			continue
+		}
+		if nsMap, ok := expectedPolicies[policy.Namespace]; ok {
+			nsMap[policy.Name] = true
+		} else {
+			expectedPolicies[policy.Namespace] = map[string]bool{
+				policy.Name: true,
+			}
+		}
+	}
+
+	return bnc.syncNetworkPoliciesCommon(expectedPolicies)
+}
+
+func (bnc *BaseNetworkController) shouldApplyMultiPolicy(mpolicy *multinetworkpolicyapi.MultiNetworkPolicy) bool {
+	policyForAnnot, ok := mpolicy.Annotations[PolicyForAnnotation]
+	if !ok {
+		// should never happen.
+		return false
+	}
+	policyForAnnot = strings.ReplaceAll(policyForAnnot, " ", "")
+	policyForNetworks := strings.Split(policyForAnnot, ",")
+	for _, networkName := range policyForNetworks {
+		networkNamespace := mpolicy.Namespace
+		a := strings.Split(networkName, "/")
+		if len(a) > 1 {
+			networkName = a[1]
+			networkNamespace = a[0]
+		}
+		if bnc.HasNAD(util.GetNADName(networkNamespace, networkName)) {
```

Comment:
in which case when NAD is not processed yet? when ovnkube-master starts, it iterate all the existing NADs first before start all the controller of NADs (see SyncNetworkControllers()). That is to avoid the problem, which could also apply on Pods where nad is defined in their NetworkAttachmentAnnot annotation.

Reply:
This is used also the a multinet policy is added at runtime, not only on the first synchonization.

So a user could create a nad and a multinet policy at the same time. If we happen to process the multinet policy before the NAD, we check that the multinet policy does not apply and then when we process the NAD we don't go back to check which multinet policies could apply because of that NAD.

Reply:
True, the assumption is that NAD is ready before multi-policy. This seems to be a problem difficult to solve. It means that for every NAD added/deleted, we'd need to trigger an update event for every existing multi-policy.

not sure what is the best way to solve this.

Reply:
based on discussion, this support is deferred